### ISSUE TITLE
`diff`: results order are not stable across runs

### ISSUE THREAD

**Issue by jqnatividad** [jqnatividad/qsv#714]
Hi @janriemer,
I just created the first test for `diff` and I had to add a `sort` as the results order are not stable across runs - maybe because of `csv-diff` multi-threading?

https://github.com/jqnatividad/qsv/blob/master/tests/test_diff.rs

Perhaps, the results should be listed in --primary-key-idx order?

**Comment by janriemer** [comment]
Ah, thanks, I totally forgot about that. Yes, it is because of multi-threading. Sorry for the inconvenience. :disappointed: 

## Initial solution to the problem

There is an easy workaround for it. `csv-diff` is able to [__sort the diffresult by line__](https://docs.rs/csv-diff/0.1.0-beta.1/csv_diff/diff_result/struct.DiffByteRecords.html#method.sort_by_line) (:warning: docs still need improvement). For modified rows, where we have lines from __both the left and the right CSV__, it will first consider the line on the left for sorting. However, modified rows will always "stick together", even when they are far apart regarding their line number.

:warning: However, sorting requires to load the whole diffresult into memory.

For sorting, we need to change [these lines](https://github.com/jqnatividad/qsv/blob/master/src/cmd/diff.rs#L87-L89) into the following (I haven't compiled this):

```diff
+ use csv_diff::diff_result::DiffByteRecords;

- let diff_byte_records_iter = csv_diff.diff(csv_rdr_left.into(), csv_rdr_right.into());

// note that this collects into a vec - naming it `_into()` was probably not the best decision by me and will probably change in the future :-/
+ let mut diff_byte_records: DiffByteRecords = csv_diff.diff(csv_rdr_left.into(), csv_rdr_right.into()).try_into()?;
+ diff_byte_records.sort_by_line();

- Ok(csv_diff_writer.write_diff_byte_records(diff_byte_records_iter)?)
+ Ok(csv_diff_writer.write_diff_byte_records(diff_byte_records.into_iter())?)
```
You can assign this issue to me. Or, if you want to try it, it is also fine with me. :slightly_smiling_face: Just let me know, which is more suitable for you.

## Future improvments
I really like your idea of sorting it by --primary-key-idx. :+1: I'll create an issue in `csv-diff` for it.

Also, there is a technique for keeping the sort order of the two CSVs while still being multi-threaded.
See this excellent resource here => [Go Concurrency Pattern in Rust - Fan In Restoring Sequence](https://github.com/munckymagik/go-concurrency-patterns-in-rust/blob/master/src/bin/08_2_fan_in_restoring_sequence.rs)

I'll investigate, whether that is something we can do in `csv-diff` (will probably only work for modified rows, as we only know added and deleted rows at the end after we have processed all rows).

**Comment by jqnatividad** [comment]
Hhhmmm... if sorting requires loading the entire file into memory, perhaps it's best to make `--sort` a `diff` option rather than the default - similar to `stats` which works in "streaming" mode by default, unless you use specify a statistic that requires loading the entire file into memory.

Being able to operate in "streaming" mode is one of the main benefits of `qsv`, so I'd like to maintain that as much as possible.

Also, qsv users can always pipe `diff` results to `sort`, like what I did in the test. We just need to mention it in `diff`'s usage text.

WDYT?

As for making the necessary changes - I'll just assign it to you since `diff` is your baby 😉

**Comment by janriemer** [comment]
@jqnatividad 

> if sorting requires loading the entire file into memory

Sorry, I probably wasn't clear enough (or maybe I misunderstood your statement):
- sorting only requires __the diff result__ itself to be loaded completely into memory
- if a diff result is empty, sorting is basically like streaming
- in other words: CSV's to be diffed (left and right) are __always__ streamed, only __sorted__ diffresult needs to allocate all necessary memory

But anyways, I completely agree with you - I'd also like to retain the "streaming" mode due to it's efficiency regarding memory usage for the diff result.

## Changes required for this issue
So, have I understood you correctly that the following change is needed now?

- [ ] Describe in the usage text of `diff` that the order of the resulting diff regarding is non-deterministic and can vary between runs

Thank you for assigning me. I'm happy to make any change that's needed, so that we get this in the hands of us nerds. :smile:

**Comment by jqnatividad** [comment]
Thanks for the clarification regarding memory usage. 

If it's just the diffresult, I think the sensible default is to sort it.

Perhaps, you can also implement sorting by the primary key and renaming the option at the same time.

**Comment by janriemer** [comment]
Great, I'll implement it and do the renaming of `--primary-key-idx` to `--key-idx` (note: we deliberately call it `-idx`, because we might have a variant in the future that works with __header names__ => e.g. `--key-headers`). :slightly_smiling_face: 

Should the default be _sort by line_ or _sort by key_ then?

Personally, I prefer _sort by line_ as default, because the output will have similarities to the input CSVs regarding ordering of records, so __the shape of the result will be more familiar to the eye.__

In that case, we would have a new option `--sort-by-key` (if I'm not mistaken, we don't need a distinction between `-idx`, or later, `-headers` here, see hint at the top).

What do you think?

**Comment by jqnatividad** [comment]
Let's go with your preference - _sort by line_ as the default, with the option to `--sort-by-key`.

And love that you're thinking about the future with `--key-headers`! 💯

**Comment by janriemer** [comment]
> FYI, I'm planning to release 0.82.0 tomorrow. Just a heads-up, doesn't mean you need to finish by tomorrow

Thank you for letting me know. I'll probably have a PR ready on Wednesday or so, so our newly planned changes probably won't make it into the new release, sorry. :worried: 

> I just want some folks to start trying out diff and giving us early feedback...

Yes, I really look forward to it! Happy releasing! :rocket:

**Comment by janriemer** [comment]
Hi @jqnatividad 

short update: unfortunately, I can't make it until today (Wednesday). End of week is probably more realistic (because weekend :slightly_smiling_face:).

As it turns out, it is not that simple to have a method `sort_by_key` on `csv_diff::DiffByteRecords`.
__But__ having a method `sort_by_columns` is probably way simpler and more flexible and should be suitable here as well.

In short: we have an interface challenge, instead of an algorithmic challenge here. :wink:    
But it's all solvable.

For more details, please see the [corresponding issue in `csv-diff`](https://gitlab.com/janriemer/csv-diff/-/issues/10).

-------------

_Off topic: Thank you for mentioning me in the latest release notes (0.82) and for quoting the csv-diff slogan ("ludicrously fast"). This has made me very happy_ :smile: :heart: :kissing_heart:

**Comment by janriemer** [comment]
Hey @jqnatividad :wave: 

just a short update on the current progress of this issue:

- Made a lot of progress in `csv-diff` for the method `sort_by_column` => tests are passing (but need some more)
- [An issue has come up](https://gitlab.com/janriemer/csv-diff/-/issues/14) with rows of type `DiffByteRecord::Modify`, where it is not clear on how to sort those, as it contains __two rows that are attached to each other__ and can't be sorted independently from each other

### Example that shows the issue
We'd like to sort the following diffresult by column _col2_
```
diffresult,col1,col2,col3
-,1,b,x    <-- DiffByteRecord::Delete
-,2,d,x    <-- DiffByteRecord::Modify's `delete` row
+,2,a,x    <-- DiffByteRecord::Modify's `add` row
+,3,c,x    <-- DiffByteRecord::Add
```
This should lead to the following sorted result:
```
diffresult,col1,col2,col3
+,2,a,x    <-- DiffByteRecord::Modify's `add` row
-,1,b,x    <-- DiffByteRecord::Delete
+,3,c,x    <-- DiffByteRecord::Add
-,2,d,x    <-- DiffByteRecord::Modify's `delete` row
```
Note that it is perfectly sorted, __but__ the `DiffByteRecord::Modify`'s `add` and `delete` row are now __separate__ and __far apart__. :warning: __This is not possible with the current data structure in `csv-diff`.__

### Possible solution (although not ideal)
- Always compare against `DiffByteRecord::Modify`'s `delete` row.
   - if the result is equal, compare against `DiffByteRecord::Modify`'s `add` row
- When comparing two `DiffByteRecord::Modify` rows with each other
   - compare `DiffByteRecord::Modify`'s `delete` row from the first with the `DiffByteRecord::Modify`'s `delete` row from the other
      - if the result is equal, compare instead both `DiffByteRecord::Modify`'s `add` rows with each other

I think this is a reasonable strategy for our first version. IMO, the most important is that it is __deterministic__.
And as I've linked above, [there is already an issue in `csv-diff`](https://gitlab.com/janriemer/csv-diff/-/issues/14), that, when implemented, will also allow the first version above (though, implementation will be more effort until we get there).

What are your opinions on this? :slightly_smiling_face:

**Comment by jqnatividad** [comment]
Hi @janriemer , thanks for the update!

As for the approach, I'll go with what you think is best.  And though being deterministic is nice, it's not necessarily a must-have.

That is, I don't want you to make decisions to make it deterministic at the cost of the utility of `diff` being reduced and/or the complexity of the csv-diff algorithm unnecessarily increased.

IMHO, having ludicrously fast csv-diffing is an AWESOME thing. So long as the diffresult is usable and more importantly, machine-readable - maybe we can do the diffresult formatting post-diffing.

BTW, I recently added a new command - `extdedup` - which uses a memory-buffered on-disk hash table using the [odht](https://crates.io/crates/odht) crate.

I needed to add it as we needed to dedup multi-gigabyte CSV files that didn't fit in memory and the existing `dedup` command was running into OOM conditions.

You may want to check out odht as I see you're using hashmaps and I can imagine `diff` being used to compare multi-gigabyte CSVs that won't fit in memory.

**Comment by janriemer** [comment]
Hey @jqnatividad,

so in our first iteration of this feature, the sorted result will look something like this (sorted by _col2_, based on the example above):

```
diffresult,col1,col2,col3
-,1,b,x    <-- DiffByteRecord::Delete
+,3,c,x    <-- DiffByteRecord::Add
-,2,d,x    <-- DiffByteRecord::Modify's `delete` row => comparison took place against this row; "d" is the largest value, therefore it has been sorted to the "last" position (it is not actually the last position, because it has a DiffByteRecord::Modify's `add` row attached to it, see next comment)
+,2,a,x    <-- DiffByteRecord::Modify's `add` row => this row hasn't been looked at during sorting, because it is attached to the above DiffByteRecord::Modify's `delete` row (they always "stick together")
```

Do you think this is acceptable for our initial release?

Now regarding your comments:

> [...] at the cost of the utility of diff being reduced and/or the complexity of the csv-diff algorithm unnecessarily increased.
#### For the first (simple) version above
__No__ change occurs to the diffing algorithm itself. The complexity is local to the (new) `sort_by_column` method.

#### For the more advanced version, where columns are perfectly sorted (and `Modify` rows detached from one another)
The diff algorithm itself will _very likely_ __not__ need to be changed. Only the internal structure of `DiffByteRecords` will have to change.   
I'm not sure about the API yet, but you could imagine something like the following that, after being called, allows the `add` and `delete` part of `DiffByteRecord::Modify` to be moved freely around (they don't "stick together") anymore.

```rust
let mut diff_records_detached: DiffByteRecordsDetached = diff_byte_records.detach();
diff_records_detached.sort_by_column(&[0, 1]);
```
Basically the call to `detach` will replace a `DiffByteRecord::Modify` row (that holds `add` and `delete` rows) with two separate rows `DiffByteRecord::ModifyAdd` and `DiffByteRecord::ModifyDelete`.   
To not make the initial implementation too complex, after calling `.detach`, a `DiffByteRecord::ModifyAdd` row will have lost relation to their corresponding `DiffByteRecord::ModifyDelete` row (and vice versa).

We would then get the perfectly sorted result (with `Modify` rows potentially far apart from another):
```
diffresult,col1,col2,col3
+,2,a,x    <-- DiffByteRecord::Modify's `add` row
-,1,b,x    <-- DiffByteRecord::Delete
+,3,c,x    <-- DiffByteRecord::Add
-,2,d,x    <-- DiffByteRecord::Modify's `delete` row
```

> [...] maybe we can do the diffresult formatting post-diffing?

Yes, of course this is a possible solution. But do you remember what we've said at the beginning of our collaboration? _Integrating `csv-diff` into `qsv` would most certainly reveal shortcomings in `csv-diff` and would be an opportunity to make the library better._ __We are now exactly at this point!__ And this is beautiful and I'm very thankful for it. After all, _sorting by column_ was _your_ idea and I think it would be a great addition to `csv-diff` itself.👍💚

> having ludicrously fast 🚀 csv-diffing is an AWESOME addition to qsv [...]

Thank you so much!🙏 I really appreciate it that csv-diffing is useful to this project.🙂

> [...] So long as the diffresult is usable and more importantly, machine-readable.

Yes, no matter our sorting strategy outlined above, the format itself won't change and will be machine-readable. Also future integration of other formats (like your proposed #733) won't be impacted.

> You may want to check out odht as I see you're using hashmaps and I can imagine diff being used to compare multi-gigabyte CSVs that won't fit in memory.

Oh, that is really cool! Thank you for the pointer. 👍 I'll evaluate it's integration in `csv-diff` and will look at how you've implemented the new `extdedup` command. Exciting stuff! 🤓
